### PR TITLE
Fixes for Python 3 migration

### PR DESCRIPTION
**Pull request by CastagnaIT**
### Check if this PR fulfills these requirements:
* [X] My changes respect the [Kodi add-on development rules](https://kodi.wiki/view/Add-on_rules)
* [X] I have read the [**CONTRIBUTING**](Contributing.md) document.
* [X] I made sure there wasn't another one [Pull Requests](../../pulls) opened for the same update/change
* [X] I have successfully tested my changes locally

#### Types of changes
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Feature change (non-breaking change which change behaviour of an existing functionality)
- [X] Improvement (non-breaking change which improve functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Description

Porting of the PR #198 
Fixes for migration to python 3, this include also fixes to allow execution with the both versions of Python:
Kodi 18.x >> Python 2
Kodi 19.x >> Python 3

Tests:
All tests are performed with a spanish profile in order to test unicode special chars cases
in follow devices:
Kodi 19.x on LibreELEC (at now the only build ready with Py3 and reuselanguageinvoker)
Kodi 18.x on Windows

List of tests:
✔Addon start
✔Cache
✔Cookies
✔Loading lists
✔Play videos
✔Library (export/remove/update context menu)
✔Auto-update all settings features
✔NFO files
✔Shared library MySQL
✔In general test all settings options
✔UpNext integration
...Other?

Who wants to give a concrete help means to verify continuously the output of the logs for every operation carried out so not only what you see on screen

### CODE REVIEW

**Comment by dagwieers** [comment]
Why are you opening a new PR? You have rights to modify my PR, so I don't see why you would duplicate the work and make it impossible for me to make changes to this PR?

**Comment by CastagnaIT** [comment]
I had to make the corrections you didn't make

**Comment by CastagnaIT** [comment]
commits are temporary

**Comment by CastagnaIT** [comment]
you said you didn't want responsibility to break the addon so i thought I'd do it this way...

**Comment by dagwieers** [comment]
Well, first of all the current code was not reviewed.

Secondly, you had three better options:
1. Start a branch from my branch, and then merge both
2. Add commits to the existing branch
3. Propose pulls requests to my branch

Obviously I cannot fix what I cannot test, but that doesn't mean my PR should not have a review, or have changes discussed. I had some more changes in the pipeline for the existing PR.

**Comment by CastagnaIT** [comment]
so wait, I'll try to join

**Comment by dagwieers** [comment]
There are quite some things we have to discuss if we want to do this properly going forward.
And the most important one is adding unit tests for everything related to things we cannot test in Travis (i.e. MySQL etc.)

**Comment by dagwieers** [comment]
You know what, I'll close my PR. Since you already have been adding more commits to this PR it is probably going to be a lot easier.

**Comment by CastagnaIT** [comment]
I was doing the pull on your branch ... that's why it went in error... ok as you prefer

**Comment by CastagnaIT** [comment]
how do run integrity tests under windows?

**Comment by dagwieers** [comment]
Look in the Makefile for the right commands. In Linux you can just do `make test` or `make unit`.
But on Windows you have to run the actual commands manually.

**Comment by dagwieers** [comment]
BTW I still need to add routing integration tests, but I added the `test/run.py` so you can start the service in the background and then run the plugin with different routes to see how it behaves.

**Comment by CastagnaIT** [comment]
I'd like to make a release, do you think it's better to wait and include this PR or can i proceed?

**Comment by Smeulf** [comment]
I have to check for MySQL again with some specific use cases if you don't mind, but I think I can do it before the evening.

**Comment by dagwieers** [comment]
As long as everything has been tested on Python 2.7 I think we're good. I haven't seen any issues (run-time or visual defects) on Python 2.7 with my PR during end-to-end testing. But we currently do not test all code-paths as we should so there's no real guarantee. (And there wasn't a guarantee before either ;-))

**Comment by CastagnaIT** [comment]
I only did a quick test with mysql...a more thorough one is appreciated
so well at least we're going the right way 😊 then to now i put it as WIP

**Comment by dagwieers** [comment]
@CastagnaIT I wasn't talking about MySQL specifically, there is a lot more code-paths we don't cover (see coverage results). And we don't have proper tests for most of the input/output functionality in Kodi/python to know for sure that we handle everything correctly wrt. unicode.

I am struggling with this for other add-ons as well, but it shouldn't be worse than before we added tests ;-)

File: resources/lib/common/kodiops.py [hunk 1]
```diff
@@ -37,7 +40,7 @@ def json_rpc(method, params=None):
                     'params': params or {}}
     request = json.dumps(request_data)
     debug('Executing JSON-RPC: {}'.format(request))
-    raw_response = unicode(xbmc.executeJSONRPC(request), 'utf-8')
+    raw_response = text(xbmc.executeJSONRPC(request)).encode('utf-8')
```
Comment (Smeulf) [inline resources/lib/common/kodiops.py]: With the series "3%" exported, I can't start an episode. I always get a  UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1013: ordinal not in range(128). This is not related to any accentuated chars, it does the same with series with no accent or anything special in the name of the series or the episode.

I don't know how to fix it, and I fear all text() functions are suffering the issue :/

See here for the first error in json_rpc() https://paste.kodi.tv/vitenogebu.kodi
And here for the same kind of error in _decrypt_chunks() https://paste.kodi.tv/olinafikon.kodi

Till this error exists, I can't go further. I'm sorry I can't help more.

Comment (dagwieers) [reply resources/lib/common/kodiops.py]: @Smeulf You have an actual Kodi with Python 3 support that works?

Comment (Smeulf) [reply resources/lib/common/kodiops.py]: No it's Python 2 (Kodi 18.2, "prod" Raspberry Pi)

Comment (Smeulf) [reply resources/lib/common/kodiops.py]: But that's not the exact same code you proposed in your PR, your's was `raw_response = text(xbmc.executeJSONRPC(request), 'utf-8')`

Comment (CastagnaIT) [reply resources/lib/common/kodiops.py]: yes is not same becouse the  `text(xbmc.executeJSONRPC(request), 'utf-8')` do not work with py2 str do not have parameters...

**Comment by dagwieers** [comment]
Using `text()` was just a hack, I think we need to move to dedicated functions. Elsewhere we use this:
```python
def to_unicode(text, encoding='utf-8'):
    ''' Force text to unicode '''
    return text.decode(encoding) if isinstance(text, bytes) else text


def from_unicode(text, encoding='utf-8'):
    ''' Force unicode to text '''
    import sys
    if sys.version_info.major == 2 and isinstance(text, unicode):  # noqa: F821; pylint: disable=undefined-variable
        return text.encode(encoding)
    return text
```

I was planning to implement this, before my branch was hijacked ;-)

File: resources/lib/common/kodiops.py (hunk 1 above)
Comment (dagwieers) [reply resources/lib/common/kodiops.py]: This works fine on Python 2:
```python
from builtins import str as text
print("%r" % text(u"wieërs", "utf-8"))
```
Returns `'wie\xc3\xabrs'`

Comment (CastagnaIT) [reply resources/lib/common/kodiops.py]: text is only a reference to str (in py2) look to the imports

**Comment by CastagnaIT** [comment]
I want to try again but tonight i can't

**Comment by CastagnaIT** [comment]
now it should work,
"unicode" is used by py2
"str" in py3 is same unicode

**Comment by CastagnaIT** [comment]
I have installed WSL Ubuntu with python and moudule requirements
with make unit/test i have only this output, it seems strange...
image

**Comment by MediaBrasil** [comment]
Kodi now support python 3.7, was merged yesterday. 

https://github.com/xbmc/xbmc/pull/16116

**Comment by CastagnaIT** [comment]
i know but there is other changes to do

**Comment by CastagnaIT** [comment]
@dagwieers 
I've been stuck for over an hour with an Kodi error
`
2019-10-15 17:32:05.658 T:42600   ERROR: XFILE::CFile::Rename - Error renaming file C:\Users\gotta\AppData\Roaming\Kodi\addons\plugin.video.netflix
2019-10-15 17:32:05.659 T:42600   ERROR: Failed to move old addon files from 'C:\Users\gotta\AppData\Roaming\Kodi\addons\plugin.video.netflix' to 'C:\Users\gotta\AppData\Roaming\Kodi\addons\temp\15c7cdf3-a9e3-4218-b4e5-ee76aafa728f'
`
every time i try install the addon does not install from error. this starting from a clean installation of Kodi on Windows 10
I have tried latest 3 days nightly matrix and leila, nothing....
I don't know what to do anymore, it never happened to me this problem
[kodi_w10_py3.log](https://github.com/CastagnaIT/plugin.video.netflix/files/3730366/kodi_w10_py3.log)

Has this ever happened to you? or kodi developers have corrupted some library?

**Comment by CastagnaIT** [comment]
well today i reinstalled again, it installed all correctly, something they did...

**Comment by dagwieers** [comment]
@CastagnaIT I don't do tests against development versions of Kodi. There's just not enough time in one day :-/

**Comment by CastagnaIT** [comment]
i'm stucked with read and save cookies with pickle
I can't get it to work in any way with py3...
I'm not interested in reading files saved by py2, it's enough that it works

the code is in: resources/lib/common/cookies.py

someone can help me how fix?
testing can be done with latest LibreELEC testbuilds
that have py3 working

**Comment by mackal** [comment]
I think you want `readBytes` instead of just `read` on line 57. Don't think encoding should be needed there either, but I'm not 100% sure.

**Comment by CastagnaIT** [comment]
thank you i believe it's working now

**Comment by CastagnaIT** [comment]
yes because peak3d has put a pull request to ruuk repository, it's a temporany fix

**Comment by chris42** [comment]
Tested this as well on latest libreelec build. After using peak3ds signals version and a few restarts it works and I am not receiving any errors yet.

**Comment by MediaBrasil** [comment]
Anyone know where we get script.module.pycryptodome updated to python 3 (for use with windows)

**Comment by CastagnaIT** [comment]
I'm not aware of which sources are for windows
but peak3d had made an attempt to get it to work:
https://forum.kodi.tv/showthread.php?tid=348495

**Comment by MediaBrasil** [comment]
I found this
https://github.com/Legrandin/pycryptodome

**Comment by CastagnaIT** [comment]
> I found this
> https://github.com/Legrandin/pycryptodome

I think it's more a problem to build the libffi library
I found something similar on a report, reported here:
https://forum.kodi.tv/showthread.php?tid=348495

but i do not understand well the building process, maybe you understand it?

**Comment by CastagnaIT** [comment]
I plan the merge to master in one/two days
if anyone has to communicate something do it now
any improvements on the code if necessary will be made on other PRs

**Comment by notoco** [comment]
Something does not work this version - at first a problem with Polish characters (ć), then addonsignals timeout
https://hastebin.com/abanamoyig.sql

The stable version obviously works.
Checked both on the old netflix profile in kodi, as well as on a completely new one

**Comment by CastagnaIT** [comment]
Can you try this?
[plugin.video.netflix_0.15.8_20191112.zip](https://github.com/CastagnaIT/plugin.video.netflix/files/3835334/plugin.video.netflix_0.15.8_20191112.zip)

**Comment by notoco** [comment]
This version works

**Pull request merged**